# Outline

1. Important Methods 
    - shape, info, isnull().sum(), describe(include = ["bool", "object"])
    - value_counts
2. Predict Chrun 

#### Syntax Formats:

1) Boolean Indexing:

2) Normalize value_counts

3) map(d)

4) Grouping


In [38]:
import pandas as pd
import numpy as np

In [39]:
df = pd.read_csv('../Data/telecom_churn.csv')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [40]:
df.shape

(3333, 20)

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null float64
Total eve minutes         3333 non-null float64
Total eve calls           3333 non-null int64
Total eve charge          3333 non-null float64
Total night minutes       3333 non-null float64
Total night calls         3333 non-null int64
Total night charge        3333 non-null float64
Total intl minutes        3333 non-null float64
Total intl calls          3333 non-null int64
Total intl charge         3333 non-null float64
Customer service calls    3333 non-null int64


bool, int64, float64 and object are the data types of our features. We see that one feature is logical (bool), 3 features are of type object, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with shape.

In [42]:
df.isnull().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

In [43]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [44]:
df.describe(include=['object','bool'])

State International plan Voice mail plan  Churn
count   3333               3333            3333   3333
unique    51                  2               2      2
top       WV                 No              No  False
freq     106               3010            2411   2850

In [45]:
df['Churn'].value_counts()

False    2850
True      483
Name: Churn, dtype: int64

In [46]:
df['International plan'].value_counts()

No     3010
Yes     323
Name: International plan, dtype: int64

In [47]:
df['Churn'].value_counts(normalize = True) 
# With normalize set to True, returns the relative frequency by dividing all values by the sum of values.

False    0.855086
True     0.144914
Name: Churn, dtype: float64

In [48]:
df['Customer service calls'].value_counts(normalize = True) 


1    0.354335
2    0.227723
0    0.209121
3    0.128713
4    0.049805
5    0.019802
6    0.006601
7    0.002700
9    0.000600
8    0.000600
Name: Customer service calls, dtype: float64

In [49]:
first_check = ['Churn', 'Customer service calls']
df[first_check].corr() # Not a strong correlation

Churn  Customer service calls
Churn                   1.00000                 0.20875
Customer service calls  0.20875                 1.00000

### Indexing and retrieving data

**what is the proportion of churned users in our dataframe?**

In [50]:
df['Churn'].mean()

0.14491449144914492

In [51]:
df['Churn'].value_counts(normalize = True)

False    0.855086
True     0.144914
Name: Churn, dtype: float64

%14.5 Customers are Churning

**What are average values of numerical features for churned users?**

Boolean indexing with one column is also very convenient. The syntax is df[P(df['Name'])], where P is some logical condition that is checked for each element of the Name column. The result of such indexing is the DataFrame consisting only of rows that satisfy the P condition on the Name column.

In [52]:
df[df['Churn'] == 1].mean()

Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

**How much time (on average) do churned users spend on the phone during daytime?**

In [53]:
df[df['Churn']==1]['Total day minutes'].mean()

206.91407867494814

**What is the maximum length of international calls among loyal users (Churn == 0) who do not have an international plan?**

In [54]:
df[(df['Churn']==0)&(df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

In [55]:
df['Churn'].unique()

array([False,  True])

In [57]:
df['International plan'].unique()

array(['No', 'Yes'], dtype=object)

### Applying Functions to Cells, Columns and Rows

df['International plan '] is an object with Yes and No values. 

**What if, we want to convert this object to Bool. Yes to be True and NO to be False**

The map method can be used to replace values in a column by passing a dictionary of the form {old_value: new_value} as its argument:

In [59]:
d = {'No': False, 'Yes': True}
df['International plan'] = df['International plan'].map(d)

In [60]:
df['International plan'].head()

0    False
1    False
2    False
3     True
4     True
Name: International plan, dtype: bool

In [61]:
df['Voice mail plan'] = df['Voice mail plan'].map(d)

In [62]:
df['Voice mail plan'].head()

0     True
1     True
2    False
3    False
4    False
Name: Voice mail plan, dtype: bool

In [64]:
second_check = ['Churn', 'Voice mail plan', 'International plan', 'Customer service calls']
df[second_check].corr() # Not a strong correlation

Churn  Voice mail plan  International plan  \
Churn                   1.000000        -0.102148            0.259852   
Voice mail plan        -0.102148         1.000000            0.006006   
International plan      0.259852         0.006006            1.000000   
Customer service calls  0.208750        -0.017824           -0.024522   

                        Customer service calls  
Churn                                 0.208750  
Voice mail plan                      -0.017824  
International plan                   -0.024522  
Customer service calls                1.000000

### Grouping

df.groupby('column name')[show these columns].function()

In [65]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 
                   'Total night minutes']
df.groupby('Churn')[columns_to_show].agg([np.mean, np.std])

Total day minutes            Total eve minutes             \
                   mean        std              mean        std   
Churn                                                             
False        175.175754  50.181655        199.043298  50.292175   
True         206.914079  68.997792        212.410145  51.728910   

      Total night minutes             
                     mean        std  
Churn                                 
False          200.133193  51.105032  
True           205.231677  47.132825

### Summary Tables

In [66]:
pd.crosstab(df['Churn'], df['International plan'])

International plan  False  True 
Churn                           
False                2664    186
True                  346    137

In [69]:
pd.crosstab(df['Churn'], df['International plan'], normalize=True)

International plan    False     True 
Churn                                
False               0.79928  0.055806
True                0.10381  0.041104

##### Pivot Table
- values – a list of variables to calculate statistics for,
- index – a list of variables to group data by,
- aggfunc – what statistics we need to calculate for groups, ex. sum, mean, maximum, minimum or something else.

In [70]:
df.pivot_table(['Total day calls', 'Total eve calls', 'Total night calls'],
               ['Churn'], aggfunc='mean')

Total day calls  Total eve calls  Total night calls
Churn                                                     
False       100.283158       100.038596         100.058246
True        101.335404       100.561077         100.399586

### DataFrame transformations

In [72]:
df['total_calls'] = df['Total day calls'] + df['Total eve calls'] + \
                df['Total night calls'] + df['Total intl calls']
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  ...  Total eve charge  \
0             45.07              197.4  ...             16.78   
1             27.47              195.5  ...             16.62   
2             41.38              121.2  ...             10.30   
3             50.90               61.9  ...              5.26   
4             28.34              148.3  ...             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  total_calls  
0                       1  False          303  
1                       1  False          332  
2                       0  False          333  
3                       2  False          255  
4                       3  False          359  

[5 rows x 21 columns]

In [73]:
df['% of Day Calls'] = df['Total day calls'] / df['total_calls']
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  ...  Total night minutes  \
0             45.07              197.4  ...                244.7   
1             27.47              195.5  ...                254.4   
2             41.38              121.2  ...                162.6   
3             50.90               61.9  ...                196.9   
4             28.34              148.3  ...                186.9   

   Total night calls  Total night charge  Total intl minutes  \
0                 91               11.01                10.0   
1                103               11.45                13.7   
2                104                7.32                12.2   
3                 89                8.86                 6.6   
4                121                8.41                10.1   

   Total intl calls  Total intl charge  Customer service calls  Churn  \
0                 3               2.70                       1  False   
1                 3               3.70                       1  False   
2                 5               3.29                       0  False   
3                 7               1.78                       2  False   
4                 3               2.73                       3  False   

   total_calls  % of Day Calls  
0          303        0.363036  
1          332        0.370482  
2          333        0.342342  
3          255        0.278431  
4          359        0.314763  

[5 rows x 22 columns]